src/Features/Core/Portable/ConvertForToForEach/AbstractConvertForToForEachCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,504 @@
+                }
+            }
+
+            // Make sure we're starting at 0.
+            var initializerValue = semanticModel.GetConstantValue(initializer, cancellationToken);
+            if (!(initializerValue.HasValue && initializerValue.Value is 0))
+            {
+                return;
+            }
+
+            // Make sure we're incrementing by 1.
+            if (stepValueExpressionOpt != null)
+            {
+                var stepValue = semanticModel.GetConstantValue(stepValueExpressionOpt);
+                if (!(stepValue.HasValue && stepValue.Value is 1))
+                {
+                    return;
+                }
+            }
+
+            var collectionType = semanticModel.GetTypeInfo(collectionExpression, cancellationToken);
+            if (collectionType.Type == null && collectionType.Type.TypeKind == TypeKind.Error)
+            {
+                return;
+            }
+
+            var containingType = semanticModel.GetEnclosingNamedType(context.Span.Start, cancellationToken);
+            if (containingType == null)
+            {
+                return;
+            }
+
+            var ienumerableType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerable_T);
+            var ienumeratorType = semanticModel.Compilation.GetSpecialType(SpecialType.System_Collections_Generic_IEnumerator_T);
+
+            // make sure the collection can be iterated.
+            if (!TryGetIterationElementType(
+                    containingType, collectionType.Type,
+                    ienumerableType, ienumeratorType,
+                    out var iterationType))
+            {
+                return;
+            }
+
+            // If the user uses the iteration variable for any other reason, we can't convert this.
+            var bodyStatements = GetBodyStatements(forStatement);
+            foreach (var statement in bodyStatements)
+            {
+                if (iterationVariableIsUsedForMoreThanCollectionIndex(statement))
+                {
+                    return;
+                }
+            }
+
+            // Looks good.  We can convert this.
+            context.RegisterRefactoring(new MyCodeAction(GetTitle(),
+                c => ConvertForToForEachAsync(
+                    document, forStatement, iterationVariable, collectionExpression,
+                    containingType, collectionType.Type, iterationType, c)));
+
+            // local functions
+            bool iterationVariableIsUsedForMoreThanCollectionIndex(SyntaxNode current)
+            {
+                if (syntaxFacts.IsIdentifierName(current))
+                {
+                    syntaxFacts.GetNameAndArityOfSimpleName(current, out var name, out _);
+                    if (name == iterationVariable.ValueText)
+                    {
+                        // found a reference.  make sure it's only used inside something like
+                        // list[i]
```

Comment:
Do we need to check if there is a conversion from the local `i` to argument passed to the indexer?

Reply:
No.  This refactoring is pattern based.  It's saying: "Hey, i noticed you're doing the indexing pattern over something.  And that thing has a enumerator.  If you want, and you feel it's fine, you can quickly switch between the two forms."

this is not an analyzer.  It's not going and fixing up things out of site.  It's not even notifying the user it's available.  It's just a tool that is available if the user goes to the 'for' and says "i would like to convert this".  Because it takes an explicit, unsolicited, action from the user, we are ok with jsut being pattern based and doing the conversion.